Optimize collision matrix building

stevesoltys opened the pull request:
These changes reduce the startup time of Apollo by a few seconds ~~and set us up for saving a few more by adding the option to apply collision flags on-demand~~.

```
Jul 17, 2018 9:02:38 PM org.apollo.Server <init>
INFO: Starting Apollo...
Jul 17, 2018 9:02:38 PM org.apollo.Server init
INFO: Initialized Release 377.
Jul 17, 2018 9:02:39 PM org.apollo.ServiceManager startAll
INFO: Starting services...
Jul 17, 2018 9:02:47 PM org.apollo.Server bind
INFO: Ready for connections.
Jul 17, 2018 9:02:47 PM org.apollo.Server main
INFO: Starting apollo took 8719 ms.
```

Related to #328.

garyttierney commented on game/src/main/java/org/apollo/game/model/area/collision/CollisionManager.java:
```diff
@@ -25,20 +24,19 @@
+	 * A {@code HashMultimap} of region coordinates mapped to positions where the tile is part of a bridged structure.
 	 */
-	private final SortedSet<Position> bridges = new TreeSet<>(POSITION_COMPARATOR);
+	private final Multimap<RegionCoordinates, Position> bridges = HashMultimap.create();
```
Why is this now per-region?

stevesoltys replied:
I only eyeballed it, but making this per-region appeared to be slightly faster (the measure being startup time) than using a set of positions.

garyttierney replied:
Faster than `TreeSet<Position>` or `HashSet<Position>`?

stevesoltys replied:
`TreeSet<Position>`, I compared the original to this one.

garyttierney commented on game/src/main/java/org/apollo/game/model/area/collision/CollisionManager.java:
```diff
@@ -67,33 +65,55 @@ public void build(boolean rebuilding) {
 			int height = tile.getHeight();
 
-			if (bridges.contains(new Position(x, y, 1))) {
+			if (regionBridges != null && regionBridges.contains(new Position(x, y, 1))) {
```
What if `regionBridges` is null?

stevesoltys replied:
If `regionBridges` is `null` then bridges doesn't contain the position, so BAU as per the prior functionality.

stevesoltys replied:
Actually I think I can remove these checks because `Multimap`returns an empty collection.

garyttierney commented on game/src/main/java/org/apollo/game/model/area/collision/CollisionManager.java:
```diff
@@ -149,7 +173,7 @@ public void apply(CollisionUpdate update) {
 	 *
 	 * @param start The start position of the ray.
-	 * @param end The end position of the ray.
+	 * @param end   The end position of the ray.
```
Try and avoid whitespace/code style changes on code you haven't changed please.

garyttierney commented on game/src/main/java/org/apollo/game/model/area/collision/CollisionManager.java:
```diff
@@ -111,7 +131,11 @@ public void apply(CollisionUpdate update) {
+			Position bridgePosition = new Position(position.getX(), position.getY(), 1);
+			Collection<Position> bridgePositions = bridges.get(bridgePosition.getRegionCoordinates());
+
+			if (bridgePositions != null && bridgePositions.contains(bridgePosition)) {
```
Same as `regionBridges` comment above.  I think there could be some edge cases around these conditions where an object extends into an adjacent region.  Any thoughts?

stevesoltys replied:
I see what you're saying here. A solution could be to get the entities in the surrounding regions which extend into our region on line 113.

The caveat with that is that instead of ignoring flags, it will be applying flags for a region we haven't initialized yet. Maybe to fix that we could filter for positions that have the same region coordinates as the one we're applying flags to.

stevesoltys committed:
Optimize collision matrix building

stevesoltys commented:
@garyttierney I've removed the on-demand collision flag functionality for now in favor of getting the optimizations merged in. The issue you brought up with adjacent regions is valid and the solution would bring us back up to ~15 seconds. We could re-introduce that functionality later when the on-demand region building is actually being used.

I also removed the whitespace changes as requested and made `bridges` a `HashSet`.

tlf30 commented:
@stevesoltys does removing the flag not impact any currently implemented functionality? I would rather see slow load times on startup than functionality being disabled. 

Also, thank you for looking into optimizations.

stevesoltys commented:
@tlf30 nope, I was referring to changes I made previously being removed. This PR in its current state should not affect anything besides performance.

On-demand collision flag building can be added in a separate PR later on.

tlf30 commented:
@stevesoltys sweet, thank you

garyttierney commented on game/src/main/java/org/apollo/game/model/area/collision/CollisionManager.java:
```diff
@@ -25,20 +27,14 @@
 	 */
-	private static final Comparator<Position> POSITION_COMPARATOR =
-		Comparator.comparingInt(Position::getX).thenComparingInt(Position::getY).thenComparingInt(Position::getHeight);
+	private final Multimap<RegionCoordinates, Position> blocked = HashMultimap.create();
```
Do we need the Map semantics here?  I think this can also be a HashSet.

stevesoltys replied:
I think this is faster because the other option would be looping over every tile in the region and doing a `HashSet#contains` before applying the tile as blocked. With the map we know which tiles are blocked already.

garyttierney replied:
Er, right.  I missed the fact that we're actually iterating over these positions.

garyttierney commented:
One more comment, LGTM besides that.

garyttierney commented:
Thanks, merged.
